proposals/list-patterns.md

```diff
@@ -69,28 +69,23 @@ Notes:
 - If the *type* is an *array_type*, the *length_pattern_clause* is disambiguated so that `int[] [0]` would match an empty integer array.
 - All other combinations are valid, for instance `T (p0, p1) [p2] { name: p3 } v` or `T (p0, p1) [p2] { p3 } v` where each clause can be omitted.
 
-> **Open question**: Should we support all these combinations?
-
 #### Pattern compatibility
 
 A *length_pattern* is compatible with any type that is *countable* — it has an accessible property getter that returns an `int` and has the name `Length` or `Count`. If both properties are present, the former is preferred.  
 A *length_pattern* is also compatible with any type that is *enumerable* — it can be used in `foreach`.
 
-A *list_pattern* is compatible with any type that is *countable* as well as *indexable* — it has an accessible indexer that takes an `Index` or `int` argument. If both indexers are present, the former is preferred.  
+A *list_pattern* is compatible with any type that is *countable* as well as *indexable* — it has an accessible indexer that takes an `Index` as an argument or otherwise an accessible indexer with a single `int` parameter. If both indexers are present, the former is preferred.  
 A *list_pattern* is also compatible with any type that is *enumerable*.
 
-A *slice_pattern* is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method that takes two `int` arguments. If both are present, the former is preferred.  
-A *slice_pattern* without a subpattern is also compatible with any type that is *enumerable*.
+A *slice_pattern* with a subpattern is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` as an argument or otherwise an accessible `Slice` method with two `int` parameters. If both are present, the former is preferred.
+A *slice_pattern* without a subpattern is compatible with any type that is compatible with a *list_pattern*.
```

Comment:
Might need to mention discard subpattern?

Reply:
With this change, we have:

- `{..}` works even without `this[Range]` or a proper `Slice` method.
- `{.._}` requires `this[Range]`/`Slice`. The codegen, however, is permitted to omit the invocation (which we do). That is not limited to discards or the slice pattern, we skip an evaluation when the output is unused, e.g. with an unused variable etc.

Reply:
Sounds like we currently don't include unused variables here, but it's something that could be done as an optimization.

Reply:
Here's an example to demonstrate it's not limited to discard: `if (this is {IntProp: {}}){}` IntProp is not called here.

Reply:
> The codegen, however, is permitted to omit the invocation (which we do).

This is a good distinction, but I do think it is good to make this not an optimization: it gets back to making sure that behavior of something like `default(ImmutableArray<int>) is { .._ }` is well-defined.

Reply:
At the moment the behavior of `default(ImmutableArray<int>) is { .._ }` is an error. "unsupported type for slice pattern"

ImmutableArray doesn't expose a Slice method or range indexer.

Reply:
Yeah, but additions are tracked by https://github.com/dotnet/runtime/issues/22928.

Reply:
My main concern is that when a type is not sliceable, we should be able to do `{1, ..}` etc. Not every list should be required to be sliceable only to be able to skip elements in the list pattern.

Reply:
I'm fine with the behavior as spec'ed in this PR. I'd let `.._` and `..var unused` require a sliceable type. It's conceptually simpler.

For `{ ..var unused }`: during binding we don't know that `var unused` is unused.
For `{ .._ }`: @333fred , I can add a note to test plan to rediscuss if you think this should not require a sliceable type.

Reply:
>For `{ .._ }`

Just like we still need an indexable type for `{_}` that we're not going to use, `{.._}` should require a sliceable type that we may not use.  I think the part that we decide to not evaluate the method, is orthogonal to *binding rules*.